// @apollo/federation/dist/composition/validate/postComposition/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var externalUnused_1 = require("./externalUnused");
exports.externalUnused = externalUnused_1.externalUnused;
var externalMissingOnBase_1 = require("./externalMissingOnBase");
exports.externalMissingOnBase = externalMissingOnBase_1.externalMissingOnBase;
var externalTypeMismatch_1 = require("./externalTypeMismatch");
exports.externalTypeMismatch = externalTypeMismatch_1.externalTypeMismatch;
var requiresFieldsMissingExternal_1 = require("./requiresFieldsMissingExternal");
exports.requiresFieldsMissingExternal = requiresFieldsMissingExternal_1.requiresFieldsMissingExternal;
var requiresFieldsMissingOnBase_1 = require("./requiresFieldsMissingOnBase");
exports.requiresFieldsMissingOnBase = requiresFieldsMissingOnBase_1.requiresFieldsMissingOnBase;
var keyFieldsMissingOnBase_1 = require("./keyFieldsMissingOnBase");
exports.keyFieldsMissingOnBase = keyFieldsMissingOnBase_1.keyFieldsMissingOnBase;
var keyFieldsSelectInvalidType_1 = require("./keyFieldsSelectInvalidType");
exports.keyFieldsSelectInvalidType = keyFieldsSelectInvalidType_1.keyFieldsSelectInvalidType;
var providesFieldsMissingExternal_1 = require("./providesFieldsMissingExternal");
exports.providesFieldsMissingExternal = providesFieldsMissingExternal_1.providesFieldsMissingExternal;
var providesFieldsSelectInvalidType_1 = require("./providesFieldsSelectInvalidType");
exports.providesFieldsSelectInvalidType = providesFieldsSelectInvalidType_1.providesFieldsSelectInvalidType;
var providesNotOnEntity_1 = require("./providesNotOnEntity");
exports.providesNotOnEntity = providesNotOnEntity_1.providesNotOnEntity;
