// @apollo/gateway/dist/loadServicesFromRemoteEndpoint.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const node_fetch_1 = require("node-fetch");
function getServiceDefinitionsFromRemoteEndpoint({ serviceList, headers = {}, serviceSdlCache, }) {
    return __awaiter(this, void 0, void 0, function* () {
        if (!serviceList || !serviceList.length) {
            throw new Error('Tried to load services from remote endpoints but none provided');
        }
        let isNewSchema = false;
        const serviceDefinitions = (yield Promise.all(serviceList.map(({ name, url, dataSource }) => {
            if (!url) {
                throw new Error(`Tried to load schema from ${name} but no url found`);
            }
            const request = {
                query: 'query GetServiceDefinition { _service { sdl } }',
                http: {
                    url,
                    method: 'POST',
                    headers: new node_fetch_1.Headers(headers),
                },
            };
            return dataSource
                .process({ request, context: {} })
                .then(({ data, errors }) => {
                if (data && !errors) {
                    const typeDefs = data._service.sdl;
                    const previousDefinition = serviceSdlCache.get(name);
                    if (previousDefinition !== typeDefs) {
                        isNewSchema = true;
                    }
                    serviceSdlCache.set(name, typeDefs);
                    return {
                        name,
                        url,
                        typeDefs: graphql_1.parse(typeDefs),
                    };
                }
                if (errors) {
                    errors.forEach(console.error);
                }
                return false;
            })
                .catch(error => {
                console.warn(`Encountered error when loading ${name} at ${url}: ${error.message}`);
                return false;
            });
        })).then(serviceDefinitions => serviceDefinitions.filter(Boolean)));
        return isNewSchema
            ? { serviceDefinitions, isNewSchema: true }
            : { isNewSchema: false };
    });
}
exports.getServiceDefinitionsFromRemoteEndpoint = getServiceDefinitionsFromRemoteEndpoint;
