// @apollo/gateway/dist/loadServicesFromStorage.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const cachedFetcher_1 = require("./cachedFetcher");
const graphql_1 = require("graphql");
const envOverrideOperationManifest = 'APOLLO_PARTIAL_SCHEMA_BASE_URL';
const envOverrideStorageSecretBaseUrl = 'APOLLO_STORAGE_SECRET_BASE_URL';
const urlFromEnvOrDefault = (envKey, fallback) => (process.env[envKey] || fallback).replace(/\/$/, '');
const urlPartialSchemaBase = urlFromEnvOrDefault(envOverrideOperationManifest, 'https://storage.googleapis.com/engine-partial-schema-prod/');
const urlStorageSecretBase = urlFromEnvOrDefault(envOverrideStorageSecretBaseUrl, 'https://storage.googleapis.com/engine-partial-schema-prod/');
const fetcher = new cachedFetcher_1.CachedFetcher();
function getStorageSecretUrl(graphId, apiKeyHash) {
    return `${urlStorageSecretBase}/${graphId}/storage-secret/${apiKeyHash}.json`;
}
function fetchStorageSecret(graphId, apiKeyHash) {
    return __awaiter(this, void 0, void 0, function* () {
        const storageSecretUrl = getStorageSecretUrl(graphId, apiKeyHash);
        const response = yield fetcher.fetch(storageSecretUrl);
        return JSON.parse(response.result);
    });
}
function getServiceDefinitionsFromStorage({ graphId, apiKeyHash, graphVariant, federationVersion, }) {
    return __awaiter(this, void 0, void 0, function* () {
        const secret = yield fetchStorageSecret(graphId, apiKeyHash);
        if (!graphVariant) {
            graphVariant = 'current';
        }
        const baseUrl = `${urlPartialSchemaBase}/${secret}/${graphVariant}/v${federationVersion}`;
        const { isCacheHit: linkFileCacheHit, result: linkFileResult, } = yield fetchLinkFile(baseUrl);
        if (linkFileCacheHit)
            return { isNewSchema: false };
        const parsedLink = JSON.parse(linkFileResult);
        const { result: configFileResult } = yield fetcher.fetch(`${urlPartialSchemaBase}/${parsedLink.configPath}`);
        const compositionMetadata = JSON.parse(configFileResult);
        const serviceDefinitions = yield fetchServiceDefinitions(compositionMetadata.implementingServiceLocations);
        return {
            serviceDefinitions,
            compositionMetadata,
            isNewSchema: true,
        };
    });
}
exports.getServiceDefinitionsFromStorage = getServiceDefinitionsFromStorage;
function fetchLinkFile(baseUrl) {
    return __awaiter(this, void 0, void 0, function* () {
        return fetcher.fetch(`${baseUrl}/composition-config-link`);
    });
}
function fetchServiceDefinitions(implementingServices) {
    return __awaiter(this, void 0, void 0, function* () {
        const serviceDefinitionPromises = implementingServices.map(({ name, path }) => __awaiter(this, void 0, void 0, function* () {
            const serviceLocation = yield fetcher.fetch(`${urlPartialSchemaBase}/${path}`);
            const { url, partialSchemaPath } = JSON.parse(serviceLocation.result);
            const { result } = yield fetcher.fetch(`${urlPartialSchemaBase}/${partialSchemaPath}`);
            return { name, url, typeDefs: graphql_1.parse(result) };
        }));
        return Promise.all(serviceDefinitionPromises);
    });
}
